### PR TITLE
test: add gas tests

### PR DESCRIPTION
**Pull request by dvilelaf**
Adding gas tests using hardhat-gas-reporter to measure the cost of contract deployment.

# RESULTS

## Ethereum
GnosisSafeL2 + VerySimpleOffchainAggregator ≈ 500$
```
·--------------------------------------------------------|---------------------------|--------------|-----------------------------·
|                  Solc version: 0.5.16                  ·  Optimizer enabled: true  ·  Runs: 1000  ·  Block limit: 30000000 gas  │
·························································|···························|··············|······························
|  Methods                                               ·               63 gwei/gas                ·       1819.07 usd/eth       │
······························|··························|·············|·············|··············|···············|··············
|  Contract                   ·  Method                  ·  Min        ·  Max        ·  Avg         ·  # calls      ·  usd (avg)  │
······························|··························|·············|·············|··············|···············|··············
|  Deployments                                           ·                                          ·  % of limit   ·             │
·························································|·············|·············|··············|···············|··············
|  CompatibilityFallbackHandler                          ·          -  ·          -  ·      953657  ·        3.2 %  ·     109.29  │
·························································|·············|·············|··············|···············|··············
|  CreateCall                                            ·          -  ·          -  ·      232052  ·        0.8 %  ·      26.59  │
·························································|·············|·············|··············|···············|··············
|  DefaultCallbackHandler                                ·          -  ·          -  ·      406608  ·        1.4 %  ·      46.60  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafe                                            ·          -  ·          -  ·     3116666  ·       10.4 %  ·     357.17  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafeL2                                          ·          -  ·          -  ·     3277885  ·       10.9 %  ·     375.65  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafeProxyFactory                                ·          -  ·          -  ·      664240  ·        2.2 %  ·      76.12  │
·························································|·············|·············|··············|···············|··············
|  MultiSend                                             ·          -  ·          -  ·      178956  ·        0.6 %  ·      20.51  │
·························································|·············|·············|··············|···············|··············
|  MultiSendCallOnly                                     ·          -  ·          -  ·      134839  ·        0.4 %  ·      15.45  │
·························································|·············|·············|··············|···············|··············
|  SignMessageLib                                        ·          -  ·          -  ·      239330  ·        0.8 %  ·      27.43  │
·························································|·············|·············|··············|···············|··············
|  SimulateTxAccessor                                    ·          -  ·          -  ·      220433  ·        0.7 %  ·      25.26  │
·························································|·············|·············|··············|···············|··············
|  third_party/canonical-weth/contracts/WETH9.sol:WETH9  ·          -  ·          -  ·      511825  ·        1.7 %  ·      58.66  │
·························································|·············|·············|··············|···············|··············
|  VerySimpleOffchainAggregator                          ·          -  ·          -  ·     1020657  ·        3.4 %  ·     116.97  │
·--------------------------------------------------------|-------------|-------------|--------------|---------------|-------------·
```
## Avalanche
GnosisSafeL2 + VerySimpleOffchainAggregator ≈ 3$
```
·--------------------------------------------------------|---------------------------|--------------|-----------------------------·
|                  Solc version: 0.5.16                  ·  Optimizer enabled: true  ·  Runs: 1000  ·  Block limit: 30000000 gas  │
·························································|···························|··············|······························
|  Methods                                               ·               25 gwei/gas                ·       24.67 usd/avax        │
······························|··························|·············|·············|··············|···············|··············
|  Contract                   ·  Method                  ·  Min        ·  Max        ·  Avg         ·  # calls      ·  usd (avg)  │
······························|··························|·············|·············|··············|···············|··············
|  Deployments                                           ·                                          ·  % of limit   ·             │
·························································|·············|·············|··············|···············|··············
|  CompatibilityFallbackHandler                          ·          -  ·          -  ·      953657  ·        3.2 %  ·       0.59  │
·························································|·············|·············|··············|···············|··············
|  CreateCall                                            ·          -  ·          -  ·      232052  ·        0.8 %  ·       0.14  │
·························································|·············|·············|··············|···············|··············
|  DefaultCallbackHandler                                ·          -  ·          -  ·      406608  ·        1.4 %  ·       0.25  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafe                                            ·          -  ·          -  ·     3116666  ·       10.4 %  ·       1.92  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafeL2                                          ·          -  ·          -  ·     3277885  ·       10.9 %  ·       2.02  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafeProxyFactory                                ·          -  ·          -  ·      664240  ·        2.2 %  ·       0.41  │
·························································|·············|·············|··············|···············|··············
|  MultiSend                                             ·          -  ·          -  ·      178956  ·        0.6 %  ·       0.11  │
·························································|·············|·············|··············|···············|··············
|  MultiSendCallOnly                                     ·          -  ·          -  ·      134839  ·        0.4 %  ·       0.08  │
·························································|·············|·············|··············|···············|··············
|  SignMessageLib                                        ·          -  ·          -  ·      239330  ·        0.8 %  ·       0.15  │
·························································|·············|·············|··············|···············|··············
|  SimulateTxAccessor                                    ·          -  ·          -  ·      220433  ·        0.7 %  ·       0.14  │
·························································|·············|·············|··············|···············|··············
|  third_party/canonical-weth/contracts/WETH9.sol:WETH9  ·          -  ·          -  ·      511825  ·        1.7 %  ·       0.32  │
·························································|·············|·············|··············|···············|··············
|  VerySimpleOffchainAggregator                          ·          -  ·          -  ·     1020657  ·        3.4 %  ·       0.63  │
·--------------------------------------------------------|-------------|-------------|--------------|---------------|-------------·
```
## Polygon
GnosisSafeL2 + VerySimpleOffchainAggregator ≈ 0.1$
```
·--------------------------------------------------------|---------------------------|--------------|-----------------------------·
|                  Solc version: 0.5.16                  ·  Optimizer enabled: true  ·  Runs: 1000  ·  Block limit: 30000000 gas  │
·························································|···························|··············|······························
|  Methods                                               ·               36 gwei/gas                ·       0.62 usd/matic        │
······························|··························|·············|·············|··············|···············|··············
|  Contract                   ·  Method                  ·  Min        ·  Max        ·  Avg         ·  # calls      ·  usd (avg)  │
······························|··························|·············|·············|··············|···············|··············
|  Deployments                                           ·                                          ·  % of limit   ·             │
·························································|·············|·············|··············|···············|··············
|  CompatibilityFallbackHandler                          ·          -  ·          -  ·      953657  ·        3.2 %  ·       0.02  │
·························································|·············|·············|··············|···············|··············
|  CreateCall                                            ·          -  ·          -  ·      232052  ·        0.8 %  ·       0.01  │
·························································|·············|·············|··············|···············|··············
|  DefaultCallbackHandler                                ·          -  ·          -  ·      406608  ·        1.4 %  ·       0.01  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafe                                            ·          -  ·          -  ·     3116666  ·       10.4 %  ·       0.07  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafeL2                                          ·          -  ·          -  ·     3277885  ·       10.9 %  ·       0.07  │
·························································|·············|·············|··············|···············|··············
|  GnosisSafeProxyFactory                                ·          -  ·          -  ·      664240  ·        2.2 %  ·       0.01  │
·························································|·············|·············|··············|···············|··············
|  MultiSend                                             ·          -  ·          -  ·      178956  ·        0.6 %  ·       0.00  │
·························································|·············|·············|··············|···············|··············
|  MultiSendCallOnly                                     ·          -  ·          -  ·      134839  ·        0.4 %  ·       0.00  │
·························································|·············|·············|··············|···············|··············
|  SignMessageLib                                        ·          -  ·          -  ·      239330  ·        0.8 %  ·       0.01  │
·························································|·············|·············|··············|···············|··············
|  SimulateTxAccessor                                    ·          -  ·          -  ·      220433  ·        0.7 %  ·       0.00  │
·························································|·············|·············|··············|···············|··············
|  third_party/canonical-weth/contracts/WETH9.sol:WETH9  ·          -  ·          -  ·      511825  ·        1.7 %  ·       0.01  │
·························································|·············|·············|··············|···············|··············
|  VerySimpleOffchainAggregator                          ·          -  ·          -  ·     1020657  ·        3.4 %  ·       0.02  │
·--------------------------------------------------------|-------------|-------------|--------------|---------------|-------------·
```

### CODE REVIEW

File: test/0_safe.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    }
+    await globals.verifyContractAddresses();
+
+    // Deploy the Gnosis safe contract
+    const Safe = await hre.ethers.getContractFactory("GnosisSafeL2");
+    const safe = await Safe.deploy();
+    await safe.deployed();
+    console.log("Safe deployed to:", safe.address);
+  });
+});
```
Comment (dvilelaf) [inline test/0_safe.js:21]: This test code comes directly from the deployment scripts.

File: test/1_oracle.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const { expect } = require("chai");
+const { ethers } = require("hardhat");
+
+describe("Oracle", function () {
+  it("Oracle aggregator deployment", async function () {
+    const Oracle = await hre.ethers.getContractFactory("VerySimpleOffchainAggregator");
+    const oracle = await Oracle.deploy(0, 120 , 18, "BTC - Valory stream", ["0x70997970C51812dc3A010C7d01b50e0d17dc79C8"]);
+    await oracle.deployed();
+  });
+});
```
Comment (dvilelaf) [inline test/1_oracle.js:10]: Got this from contracts-oracle.

**Comment by dvilelaf** [comment]
CI is failing when trying to add ethereumjs:
```
Command: git
Arguments: ls-remote --tags --heads ssh://git@github.com/ethereumjs/ethereumjs-abi.git
Directory: /home/runner/work/contracts-amm/contracts-amm
Output:
Warning: Permanently added the ECDSA host key for IP address '140.82.113.3' to the list of known hosts.
git@github.com: Permission denied (publickey).
```

File: hardhat.config.js [hunk 3]
```diff
@@ -3,6 +3,7 @@ require("solidity-coverage");
 require("hardhat-deploy");
 require("@ethersproject/constants");
 require("@gnosis.pm/safe-contracts");
+require("hardhat-gas-reporter");
```
Comment (kupermind) [inline hardhat.config.js:7]: You might want to use this one as well:
`require("hardhat-contract-sizer");`

and then:
`npx hardhat size-contracts`

Here's the link for configs, if needed:
https://www.npmjs.com/package/hardhat-contract-sizer

**Pull request closed without merging**